static/js/stream_data.js

```diff
@@ -270,8 +270,7 @@ exports.update_calculated_fields = function (sub) {
     // If stream is public then any user can subscribe. If stream is private then only
     // subscribed users can unsubscribe.
     // Guest users can't subscribe themselves to any stream.
-    sub.should_display_subscription_button = sub.subscribed ||
-        !page_params.is_guest && !sub.invite_only;
+    sub.should_display_subscription_button = sub.subscribed && !sub.invite_only;
```

Comment:
I believe the old code was correct -- if the user is subscribed, we'll show an "unsubscribe" button.